src/ray/raylet/node_manager.cc

```diff
@@ -341,15 +341,17 @@ void NodeManager::ClientAdded(const ClientTableDataT &client_data) {
   // Establish a new NodeManager connection to this GCS client.
   auto status = ConnectRemoteNodeManager(client_id, client_data.node_manager_address,
                                          client_data.node_manager_port);
-  // A disconnected client has 2 entries in the client table (one for being
-  // inserted and one for being removed). When a new raylet starts, ClientAdded
-  // will be called with the disconnected client's first entry, which will cause
-  // IOError and "Connection refused".
   if (!status.ok()) {
-    RAY_LOG(WARNING) << "Failed to connect to client " << client_id
-                     << " in ClientAdded. TcpConnect returned status: "
-                     << status.ToString() << ". This may be caused by "
-                     << "trying to connect to a node manager that has failed.";
+    // This is not a fatal error for raylet, but it should not happen.
+    // We need to broadcase this message.
+    std::string type = "raylet_connection_error";
+    std::ostringstream error_message;
+    error_message << "Failed to connect to client " << client_id
+                  << " in ClientAdded. TcpConnect returned status: " << status.ToString()
+                  << ". This error should not happen.";
```

Comment:
This can happen if there is a race condition right? Maybe we could just say "Failed to connect to ray node <client_id>. This may be since the node was recently removed."

Reply:
I have changed according to your comment.
Yes. there could be 2 cases.
1. If raylet crashes or is killed by SIGKILL and the heartbeat timeout (30 seconds) has not happened, this may happen. This could be the case that happens most since 30 seconds is very long.
2. Race condition: The new node just gets the client information and tries to connect to the existing nodes and at this time the disconnecting message arrives. This it is very rare since the time interval is very short.